more_itertools/more.py

```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
         datetime.datetime(2019, 1, 1, 0, 0)
         >>> next(items)
         datetime.datetime(2019, 1, 2, 0, 0)
 
     """
-    argc = len(args)
-    if argc == 1:
-        stop, = args
-        start = type(stop)(0)
-        step = 1
-    elif argc == 2:
-        start, stop = args
-        step = 1
-    elif argc == 3:
-        start, stop, step = args
-    else:
-        err_msg = 'numeric_range takes at most 3 arguments, got {}'
-        raise TypeError(err_msg.format(argc))
-
-    values = (start + (step * n) for n in count())
-    zero = type(step)(0)
-    if step > zero:
-        return takewhile(partial(gt, stop), values)
-    elif step < zero:
-        return takewhile(partial(lt, stop), values)
-    else:
-        raise ValueError('numeric_range arg 3 must not be zero')
+    _EMPTY_HASH = hash(range(0, 0))
+
+    def __init__(self, *args):
+        argc = len(args)
+        if argc == 1:
+            self._stop, = args
+            self._start = type(self._stop)(0)
+            self._step = 1
+        elif argc == 2:
+            self._start, self._stop = args
+            self._step = 1
+        elif argc == 3:
+            self._start, self._stop, self._step = args
+        else:
+            if argc == 0:
+                err_msg = 'numeric_range expected 1 argument, got {}'
+            else:
+                err_msg = 'numeric_range expected at most 3 arguments, got {}'
+            raise TypeError(err_msg.format(argc))
+
+        self._zero = type(self._step)(0)
+        if self._step == self._zero:
+            raise ValueError('numeric_range() arg 3 must not be zero')
+        self._growing = self._step > self._zero
+
+    def __bool__(self):
+        if self._growing:
+            return self._start < self._stop
+        else:
+            return self._start > self._stop
+
+    def __contains__(self, elem):
+        if self._growing:
+            if self._start <= elem < self._stop:
+                return (elem - self._start) % self._step == self._zero
+        else:
+            if self._start >= elem > self._stop:
+                return (self._start - elem) % (-self._step) == self._zero
+
+        return False
+
+    def __eq__(self, other):
+        return (self._start == other._start
+                and self._step == other._step
+                and self._get_by_index(-1) == other._get_by_index(-1))
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            return self._get_by_index(key)
+        elif isinstance(key, slice):
+            start = self._start if key.start is None \
+                else self._get_by_index(key.start)
+            stop = self._stop if key.stop is None \
+                else self._get_by_index(key.stop)
+            step = self._step if key.step is None else key.step * self._step
+            return numeric_range(start, stop, step)
+        else:
+            raise TypeError(
+                "numeric range indices must be integers or slices, not str")
```

Comment:
`str` is hardcoded here - this should probably check the type of what was passed in, right?

Reply:
My mistake!

Reply:
Fixed.